apps/kg/models/mxnet/score_fun.py

```diff
@@ -46,6 +57,125 @@ def fn(heads, relations, tails, num_chunks, chunk_size, neg_sample_size):
                 return gamma - nd.norm(heads - tails, ord=1, axis=-1)
             return fn
 
+class TransRScore(nn.Block):
+    def __init__(self, gamma, projection_emb, relation_dim, entity_dim):
+        super(TransRScore, self).__init__()
+        self.gamma = gamma
+        self.projection_emb = projection_emb
+        self.relation_dim = relation_dim
+        self.entity_dim = entity_dim
+
+    def edge_func(self, edges):
+        head = edges.data['head_emb']
+        tail = edges.data['tail_emb']
+        rel = edges.data['emb']
+        score = head + rel - tail
+        return {'score': self.gamma - nd.norm(score, ord=1, axis=-1)}
+
+    def prepare(self, g, gpu_id, trace=False):
+        head_ids, tail_ids = g.all_edges(order='eid')
+        projection = self.projection_emb(g.edata['id'], gpu_id, trace)
+        projection = projection.reshape(-1, self.entity_dim, self.relation_dim)
+        head_emb = g.ndata['emb'][head_ids.as_in_context(g.ndata['emb'].context)].expand_dims(axis=-2)
+        tail_emb = g.ndata['emb'][tail_ids.as_in_context(g.ndata['emb'].context)].expand_dims(axis=-2)
+        g.edata['head_emb'] = nd.batch_dot(head_emb, projection).squeeze()
+        g.edata['tail_emb'] = nd.batch_dot(tail_emb, projection).squeeze()
```

Comment:
i don't feel that we need `prepare`. we can put all computation in `forward`. It shouldn't affect the speed.

Reply:
Yes, combining forward and prepare will not affect the speed.
The reason I use prepare here is to avoid breaking the interface of original forward func and make the code more flexible.  If we want to add other score funcs, we may reuse prepare to do some data preprocessing or transformation.